File: Search in Rotated Sorted Array/Search in Rotated Sorted Array.cpp
class Solution {
public:
    int search(int A[], int n, int target) {
        int l = 0;
        int r = n - 1;
        while(l <= r){
            int m = (l + r) / 2;
            if(A[m] == target)      return m;       // Found at middle
            //if(A[l] == target)      return l;       // Found at left
            //if(A[r] == target)      return r;       // Found at right
            
            if(A[l] <= A[m]){                        // Pivot is in the right half
                if((target >= A[l]) && (target < A[m])){       // Choose left
                    r = m - 1;
                }
                else{                                           // Choose right
                    l = m + 1;
                }
            }
            else{                                   // Pivot is in the left half
                if((target > A[m]) && (target <= A[r])){       // Choose right
                    l = m + 1;
                }
                else{                                           // Choose left
                    r = m - 1;
                }
            }
        }
        
        return -1;      // Not found
    }
};

File: Search in Rotated Sorted Array/Search in Rotated Sorted Array.java
public class Solution {
    public int search(int[] A, int target) {
        int start = 0;
        int end = A.length - 1;
        
        while(start <= end){
            int mid = start + end;
            mid /= 2;
            boolean leftRotated = false;
            boolean rightRotated = false;
            if(A[start] > A[mid])            leftRotated = true;
            if(A[mid] > A[end])        rightRotated = true;
            
            if(A[mid] == target){
                return mid;
            }
            else if(A[mid] < target){
                if(leftRotated){
                    if(A[start] <= target){
                        end = mid - 1;
                    }
                    else{
                        start = mid + 1;
                    }
                }
                else{
                    start = mid + 1;
                }
            }
            else{
                if(rightRotated){
                    if(A[end] >= target){
                        start = mid + 1;
                    }
                    else{
                        end = mid - 1;
                    }
                }
                else{
                    end = mid - 1;
                }
            }
        }
        return -1;
    }
}
